lisp/ess-r-mode.el

```diff
@@ -711,84 +711,53 @@ Executed in process buffer."
-           (concat
-            "(defun R-X.Y (&optional start-args)
-  \"Call the R version 'R-X.Y' using ESS.
-This function was generated by `ess-r-versions-create'.\"
-  (interactive \"P\")
-  (let ((inferior-R-version \"R-X.Y\")
-        (inferior-ess-r-program-name \""
-            (if ess-microsoft-p "Rterm" "R") "-X.Y\"))
-    (R start-args)))
-")))
-
-      (with-current-buffer eval-buf
-        ;; clear the buffer.
-        (delete-region (point-min) (point-max))
-
-        ;; Find which versions of R we want.  Remove the pathname, leaving just
-        ;; the name of the executable.
-        (setq versions
-              (if ess-microsoft-p
-                  (mapcar (lambda(v) (ess-rterm-arch-version v 'give-cons))
-                          ess-rterm-version-paths)
-                ;;        ^^^^^^^^^^^^^^^^^^^^^^^ from ./ess-site.el at start
-                ;; else (non-MS):
-                (ess-uniq-list
-                 (mapcar 'file-name-nondirectory
-                         (apply 'nconc
-                                (mapcar 'ess-find-exec-completions
-                                        ess-r-versions))))))
-        (setq r-versions-created ; also for returning at end.
-              (if ess-microsoft-p
-                  (mapcar 'car versions)
-                versions))
-        (ess-write-to-dribble-buffer
-         (format "(R): ess-r-versions-create making M-x defuns for \n %s\n"
-                 (mapconcat 'identity r-versions-created "\n ")))
-
-        ;; Iterate over each string in VERSIONS, creating a new defun each time.
-        (while versions
-          (let* ((version (car versions))
-                 (ver (if ess-microsoft-p (car version) version))
-                 (beg (point)))
-
-            (setq versions (cdr versions))
-            (insert template)
-            (goto-char beg)
-            (while (search-forward "R-X.Y" nil t) ;; in all cases
-              (replace-match ver t t))
-            (when (and ess-microsoft-p
-                       (fboundp 'w32-short-file-name))
-              (goto-char beg)
-              (while (search-forward "Rterm-X.Y" nil t)
-                (replace-match (w32-short-file-name (cdr version)) t t)))
-            (goto-char (point-max))))
-        ;; buffer has now been created with defuns, so eval them!
-        (eval-buffer))
-      (unless (and (boundp 'ess-debugging) ess-debugging)
-        (kill-buffer eval-buf))
-
+  "Generate functions for starting other versions of R.
+See `ess-r-versions' for strings that determine which functions
+are created.  On MS Windows, this works using
+`ess-rterm-version-paths' instead.
+
+The result is a list of strings (or nil) with the names of new
+functions, if any, that were created.  The functions will
+normally be placed on the menubar and stored as
+`ess-r-versions-created' upon ESS initialization."
+  (when ess-r-versions
+    ;; try to find R versions in ess-r-versions.
+    (let (versions r-versions-created)
```

Comment:
I think we can avoid this `let` here and work with `r-versions-created` directly. We are setq-ing it anyways. 

Plus, let's rename into `ess-r-created-runners`.